File: src/dlgs/dlg_edit.rs
use std::cell::{Cell, RefCell};
use std::rc::Rc;
use winsafe::{self as w, gui, prelude::*};

use super::{Input, WndPicture, ids};
use crate::id3v2;

#[derive(Clone)]
pub struct DlgEdit {
	pub(super) wnd: gui::WindowModal,
	pub(super) inputs: Vec<Input>, // all checkbox + textbox for the fields
	pub(super) chk_pic: gui::CheckBox,
	pub(super) wnd_pic: WndPicture,
	pub(super) lbl_pic: gui::Label,
	pub(super) lst_frames: gui::ListView<id3v2::Frame>,
	pub(super) btn_uncheck_all: gui::Button,
	pub(super) btn_check_filled: gui::Button,
	pub(super) sel_tags: Rc<RefCell<Vec<id3v2::Tag>>>, // these will be modified and then returned
	pub(super) user_clicked_ok: Rc<Cell<bool>>,
}

impl DlgEdit {
	/// Creates the dialog and displays it, blocking until it's closed.
	///
	/// Takes ownership of the tags, and returns the modified tags.
	#[must_use]
	pub fn show(
		parent: &(impl GuiParent + 'static),
		sel_tags: Vec<id3v2::Tag>,
	) -> w::AnyResult<Option<Vec<id3v2::Tag>>> {
		let no_res = (gui::Horz::None, gui::Vert::None);

		let wnd = gui::WindowModal::new_dlg(ids::DLG_EDIT);
		let inputs = (ids::CHK_ARTIST..=ids::CHK_COMMENT)
			.step_by(2)
			.zip(FIELDS_NAME4.iter())
			.map(|(chk_id, name4)| Input::new(&wnd, *name4, chk_id))
			.collect::<Vec<_>>();
		let chk_picture = gui::CheckBox::new_dlg(&wnd, ids::CHK_PICTURE, no_res);
		let wnd_pic = WndPicture::new(&wnd, gui::dpi(420, 60), gui::dpi(200, 200), no_res);
		let lbl_pic = gui::Label::new_dlg(&wnd, ids::LBL_PICTURE, no_res);
		let lst_frames =
			gui::ListView::new_dlg(&wnd, ids::LST_FRAMES, no_res, Some(ids::MNU_FRAMES));
		let btn_uncheck_all = gui::Button::new_dlg(&wnd, ids::BTN_UNCHECK_ALL, no_res);
		let btn_check_filled = gui::Button::new_dlg(&wnd, ids::BTN_CHECK_FILLED, no_res);
		let sel_tags = Rc::new(RefCell::new(sel_tags));
		let user_clicked_ok = Rc::new(Cell::new(false));

		let new_self = Self {
			wnd,
			inputs,
			chk_pic: chk_picture,
			wnd_pic,
			lbl_pic,
			lst_frames,
			btn_uncheck_all,
			btn_check_filled,
			sel_tags,
			user_clicked_ok,
		};
		new_self.events();
		new_self.show_modal(parent)
	}

	fn show_modal(&self, parent: &impl GuiParent) -> w::AnyResult<Option<Vec<id3v2::Tag>>> {
		self.wnd.show_modal(parent)?;
		if self.user_clicked_ok.get() {
			let edited_tags = self.sel_tags.replace(Vec::new());
			Ok(Some(edited_tags)) // user clicked OK
		} else {
			Ok(None) // user clicked Cancel
		}
	}
}

/// Frame names for each Input, in order.
pub const FIELDS_NAME4: &[&str] = &[
	"TPE1", "TIT2", "TIT3", "TALB", "TRCK", "TYER", "TCON", "TPE3", "TPUB", "TOPE", "TOAL", "TORY",
	"TCOM", "TEXT", "COMM",
];

File: dlg/dlgedit/DlgEdit.go
//go:build windows

package dlgedit

import (
	"id3fit/dlg/ids"
	"id3fit/dlg/wndpicture"
	"id3fit/id3v2"

	"github.com/rodrigocfd/windigo/ui"
	"github.com/rodrigocfd/windigo/win/co"
)

// Modal dialog to edit the fields of one or many tags.
type DlgEdit struct {
	wnd            *ui.Modal
	inputs         []CheckInput // checkbox + input fields
	chkPick        *ui.CheckBox
	wndPic         *wndpicture.WndPicture
	lblPic         *ui.Static
	lstFrames      *ui.ListView
	btnUncheckAll  *ui.Button
	btnCheckFilled *ui.Button

	tags   []*id3v2.Tag // cloned tags to be modified
	result co.ID        // returned when the user closes the modal
}

// A pair of a checkbox + input fields.
type CheckInput struct {
	chk *ui.CheckBox
	txt ui.ChildControl // Edit or ComboBox
}

// Constructor; blocks until the modal is closed.
func ShowNew(parent ui.Parent, tags []*id3v2.Tag) co.ID {
	wnd := ui.NewModalDlg(parent, ids.DLG_EDIT)

	inputs := make([]CheckInput, 0, (ids.TXT_COMMENT-ids.CHK_ARTIST+1)/2)
	for id := ids.CHK_ARTIST; id <= ids.TXT_COMMENT; id += 2 {
		chk := ui.NewCheckBoxDlg(wnd, id, ui.LAY_NONE_NONE)
		var inp ui.ChildControl
		if id == ids.CHK_GENRE {
			inp = ui.NewComboBoxDlg(wnd, ids.CMB_GENRE, ui.LAY_NONE_NONE)
		} else {
			inp = ui.NewEditDlg(wnd, id+1, ui.LAY_NONE_NONE)
		}
		inputs = append(inputs, CheckInput{chk, inp})
	}

	chkPic := ui.NewCheckBoxDlg(wnd, ids.CHK_PICTURE, ui.LAY_NONE_NONE)
	wndPic := wndpicture.New(wnd, ui.DpiX(420), ui.DpiY(60), ui.DpiX(200), ui.DpiY(200))
	lblPic := ui.NewStaticDlg(wnd, ids.LBL_IMAGE_DESCR, ui.LAY_NONE_NONE)

	lstFrames := ui.NewListViewDlg(wnd, ids.LST_FRAMES, ids.MNU_FRAMES, ui.LAY_NONE_NONE)
	btnUncheckAll := ui.NewButtonDlg(wnd, ids.BTN_UNCHECK_ALL, ui.LAY_NONE_NONE)
	btnCheckFilled := ui.NewButtonDlg(wnd, ids.BTN_CHECK_FILLED, ui.LAY_NONE_NONE)

	me := &DlgEdit{wnd, inputs,
		chkPic, wndPic, lblPic,
		lstFrames, btnUncheckAll, btnCheckFilled,
		tags, co.ID_CANCEL}

	me.events()
	me.wnd.ShowModal() // blocks until the modal is closed
	return me.result
}

// Relationship between fields and frame IDs.
var FIELD_NAMES = map[uint16]string{
	ids.CHK_ARTIST:      "TPE1",
	ids.CHK_TITLE:       "TIT2",
	ids.CHK_SUBTITLE:    "TIT3",
	ids.CHK_ALBUM:       "TALB",
	ids.CHK_TRACK:       "TRCK",
	ids.CHK_YEAR:        "TYER",
	ids.CHK_GENRE:       "TCON",
	ids.CHK_PERFORMER:   "TPE3",
	ids.CHK_PUBLISHER:   "TPUB",
	ids.CHK_ORIG_ARTIST: "TOPE",
	ids.CHK_ORIG_ALBUM:  "TOAL",
	ids.CHK_ORIG_YEAR:   "TORY",
	ids.CHK_COMPOSER:    "TCOM",
	ids.CHK_LYRICIST:    "TEXT",
	ids.CHK_COMMENT:     "COMM",
}
